homeassistant/components/improv_ble/config_flow.py

```diff
@@ -0,0 +1,368 @@
+    _unsub: Callable[[], None] | None = None
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._device: ImprovBLEClient | None = None
+        # Populated by user step
+        self._discovered_devices: dict[str, BluetoothServiceInfoBleak] = {}
+        # Populated by bluetooth, reauth_confirm and user steps
+        self._discovery_info: BluetoothServiceInfoBleak | None = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user step to pick discovered device."""
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            address = user_input[CONF_ADDRESS]
+            await self.async_set_unique_id(address, raise_on_progress=False)
+            # Guard against the user selecting a device which has been configured by
+            # another flow.
+            self._abort_if_unique_id_configured()
+            self._discovery_info = self._discovered_devices[address]
+            return await self.async_step_start_improv()
+
+        current_addresses = self._async_current_ids()
+        for discovery in async_discovered_service_info(self.hass):
+            if (
+                discovery.address in current_addresses
+                or discovery.address in self._discovered_devices
+                or not device_filter(discovery.advertisement)
+            ):
+                continue
+            self._discovered_devices[discovery.address] = discovery
+
+        if not self._discovered_devices:
+            return self.async_abort(reason="no_devices_found")
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_ADDRESS): vol.In(
+                    {
+                        service_info.address: (
+                            f"{service_info.name} ({service_info.address})"
+                        )
+                        for service_info in self._discovered_devices.values()
+                    }
+                ),
+            }
+        )
+        return self.async_show_form(
+            step_id="user",
+            data_schema=data_schema,
+            errors=errors,
+        )
+
+    async def async_step_bluetooth(
+        self, discovery_info: BluetoothServiceInfoBleak
+    ) -> FlowResult:
+        """Handle the Bluetooth discovery step."""
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        service_data = discovery_info.service_data
+        improv_service_data = ImprovServiceData.from_bytes(
+            service_data[SERVICE_DATA_UUID]
+        )
+        if improv_service_data.state in (State.PROVISIONING, State.PROVISIONED):
+            _LOGGER.debug(
+                "Device is already provisioned: %s", improv_service_data.state
+            )
+            return self.async_abort(reason="already_provisioned")
+        self._discovery_info = discovery_info
```

Comment:
I was thinking about the missing name problem we discussed on discord. 

For a followup PR, If we get here and the name is still missing, we could do something like we do in xiaomi_ble https://github.com/home-assistant/core/blob/61104dd726157821a0bb1be5af2c4f3e4881934d/homeassistant/components/xiaomi_ble/config_flow.py#L54

Reply:
Same comment as above: The bluetooth integration should abort flows and re-trigger them when advertised data has changed.